Fix for pruning bug : issue 285

ptschip opened the pull request:
Here we use chainActive.Height() rather than pindex->nHeight.  pindex->nHeight will work but only when the chain is fully sync'd whereas using the chainActive keeps the block files updated with the correct heights which can then be used for appropriate pruning while syncing a new chain.

gandrewstone commented:
Is it correct that pindex->nHeight is the height of the block we are dealing with, while chain.Height() is the current height of the main chain?

If so, I would think that the original code is correct, since it writes the blocks height, not the chain's height at the time we happened to receive the block.  What am I misunderstanding?

ptschip commented:
pindex->nHeight refers to the blockindex and not the actual chain.  If
you actually print out the heights you will see that as you're doing IBD
the pindex->nHeight is pointing to your last entry in the blockindex,
which because we download all the header before we start IBD, is >
450000.   So the height in the block.dat file is this height, for all of
them, when you're doint IBD...so that's why the files never get delted. 
I think that part of the problem is two fold, in the past the headers
didn't get downloaded all at once before starting IBD, and secondly i
think that the AcceptBlock() function is assuming we have a sync'd chain...

the fact is this works perfectly, but now I wonder, why is acceptblock()
not using the pindex of the actual block?  there seems to something odd
about that, hmm, maybe some other bug in accept block?


On 09/02/2017 7:34 AM, gandrewstone wrote:
>
> Is it correct that pindex->nHeight is the height of the block we are
> dealing with, while chain.Height() is the current height of the main
> chain?
>
> If so, I would think that the original code is correct, since it
> writes the blocks height, not the chain's height at the time we
> happened to receive the block. What am I misunderstanding?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/291#issuecomment-278677189>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0GwbRFR8Qv7RyuCInMi6qklTciG4ks5razIFgaJpZM4L7ijj>.
>

ptschip commented:
@gandrewstone  ok i think this is what is the problem and why this solution works.  So let say you're doing IBD...you filling up your block.dat file..you're actually getting the right pindex->nHeight at that point...now what happens when a newly mined block arrives?  well it has to get stored in the block.dat file...and it then updates the nLastHeight to be the height of the newly mined block...now you're got your problem because that newly mined block is obviously not part of chainActive.

ptschip commented:
@gandrewstone  hmmm, thinking about this more, IF this is actually the case, that new blocks are coming and causing the nheight to be too large, then we still have a problem...because if we now have new blocks in a very old block.dat file, and we will end up deleting those new blocks when the block.dat file gets unlinked during pruning...so we would additionally have to prevent any new blocks from getting downloaded while we're doing IBD...or in other words, still used the old pindex->nHeight but prevent any new blocks from coming in. (if that's really the case...i'll do some more testing and maybe come up with a separate solution)

ptschip committed:
Fix pruning when syncing a chain for the first time

When  syncing a chain with pruning enabled there are
at times new blocks arriving which make the nLastHeight equal
to the tip of the blockchain, however this prevents block files
from being removed during pruning.  By not downloading new blocks
until the chain no longer in IsInitialDownload() the issue is
prevented.

Unfortunately this workaround also requires that we still allow
all blocks to be downloaded on regtest.  While the fix works fine
it's never a great idea to test essentially different code
from what is released on mainnet.  And while there  are a handful of these
cases in the code base providing a precedenche,  it would still be
good to find a more broad based fix for this issue.

ptschip committed:
Fix pruning.py

This is a workaround to the longstanding issue of "broken pipes" which
is caused by a bug in the http server.  Using rpcservertimeout=0 prevents
some of the numerous hangups we can see in long running tests.

gandrewstone commented:
> pindex->nHeight refers to the blockindex and not the actual chain. If you actually print out the heights 
> you > will see that as you're doing IBD the pindex->nHeight is pointing to your last entry in the 
> blockindex, which > because we download all the header before we start IBD, is > 450000.

Is that a bug then?  because in class CBlockIndex its defined as the height of that block not the current height of the main chain:
    //! height of the entry in the chain. The genesis block has height 0
    int nHeight;

Although from inspecting the code (AddToBlockIndex), it looks correct.  Unfortunately, if it is the case that we have recent blocks sprinkled all over  the old files due to receiving new blocks during IBD, then we need to keep those old blocks around.

This whole technique of dumping the blocks in a file is problematic.  File systems are pretty robust these days, we might as well just dump them into individual files named something like "height-hash.dat".  Or even better store them in leveldb.

But given that we can't re-architect the entire block storage to solve this bug, two possibilities are to:
1. ignore blocks that are N blocks > the current IBD position (we'll IBD them later), 
2. put out-of-order blocks in a separate file

ptschip commented:
@gandrewstone here's a safer approach and just a one liner...have a look...the only downside is once again we have to use the "regtest" workaround.

ptschip commented:
On 09/02/2017 12:10 PM, gandrewstone wrote:
>
>     pindex->nHeight refers to the blockindex and not the actual chain.
>     If you actually print out the heights
>     you > will see that as you're doing IBD the pindex->nHeight is
>     pointing to your last entry in the
>     blockindex, which > because we download all the header before we
>     start IBD, is > 450000.
>
> Is that a bug then? because in class CBlockIndex its defined as the
> height of that block not the current height of the main chain:
> //! height of the entry in the chain. The genesis block has height 0
> int nHeight;
>
> Although from inspecting the code (AddToBlockIndex), it looks correct.
> Unfortunately, if it is the case that we have recent blocks sprinkled
> all over the old files due to receiving new blocks during IBD, then we
> need to keep those old blocks around.
>
> This whole technique of dumping the blocks in a file is problematic.
> File systems are pretty robust these days, we might as well just dump
> them into individual files named something like "height-hash.dat". Or
> even better store them in leveldb.
>
> But given that we can't re-architect the entire block storage to solve
> this bug, two possibilities are to:
>
>  1. ignore blocks that are N blocks > the current IBD position (we'll
>     IBD them later),
>
I just saw your comment...yes, i think this is the way to go and is the
second fix i just pushed up.
>
>  1. put out-of-order blocks in a separate file
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/291#issuecomment-278758107>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0HiiJBYzORWEBaaC0JqABTG_iVWdks5ra3KcgaJpZM4L7ijj>.
>

ftrader commented:
The latest fix does correct the problem successfully for me, based on a test on mainnet.

ftrader commented:
Since the fix for this is still WIP, I shall add that it looks to me like the root cause of the problem was the removal of a check for proximity to the active chain tip in https://github.com/BitcoinUnlimited/BitcoinUnlimited/commit/fdf065a0d2d5b55484008a5891f51120dd0cfd10 . Specifically the commit notes about "issue blocks to req mgr whenever INVs come in" and we can see a guarding condition on  `IsChainNearlySyncd()` being removed in the `if` statement that guards the section in question.

In Bitcoin Core, this is still handled by call `CanFetchDirect()`, which seems to have a similar purpose (returns True if in proximity to the active chain tip). It is this call in Core which protects the ordering of the block data in blk files from becoming 'polluted' by incoming recent (network chain tip) blocks.

I've done a simple test on mainnet with `CanDirectFetch(chainparams.GetConsensus())` added back as a condition before calling `requester.AskFor(inv, pfrom)`, and this seems to work just as well at preventing the chain tip blocks from disturbing the pruning.

ptschip commented:
@ftrader good analysis...it looks like it was inadvertently taken out and then slipped through because the pruning tests don't check for it...i should add a test in pruning and make sure this doesn't happen again.

ptschip committed:
Add test new test for pruning.py

When syncing a new chain from the start we have to ensure
that newly mined blocks do not get into the chain until
the chain is almost sync'd otherwise block files may not
be prunable.

NOTE: This test is in place but is currently disabled until we
can find a solution to the pruning workaround for "regtest" that
is needed for the regression test scripts to run properly.

ptschip commented:
taking this out of WIP.  It's as good as we can do for now.  The python test is in place but deactivated because we need to deal with the "regtest" workaround first.  However, that has to wait until we can implement SENDHEADERS and is beyond the scope of this PR.  But the main point is that the pruning now works as intended.  The regression test can wait until later IMO.

sickpig commented:
The `pruning.py` is passing without the spurious "Broken pipe" error messages (only 2 instances just after shutting down the test nodes).

gandrewstone commented:
this is great detective work!  One issue is that I don't want to lose all of this knowledge.  What I mean is that its a real stretch to imagine that someone contemplating the line on 5267 "!IsInitialBlockDownload()" will realize that the reason is that this causes out-of-order blocks to be written to the block files which then screws up pruning.  Especially since this whole thing is caused by a pretty bad implementation of the block files...

Let's make sure we clearly comment whenever a LOC has significant and indirect ramifications.  But for expediency, I'm going to merge this and add the comment as a separate check in.

ftrader commented:
@gandrewstone : this seems to need retargeting to `release` as well. It is not included in 1.0.1.0 .
